src/arch_ppc.py

```diff
@@ -734,24 +811,90 @@ def make_memory_access(arg: Argument, size: int) -> List[Location]:
             else:
                 assert not any(isinstance(a, AsmAddressMode) for a in args)
                 inputs = [r for r in args[1:] if isinstance(r, Register)]
-        elif mnemonic in cls.instrs_implicit_destination:
+            if mnemonic.endswith("."):
+                # Instructions ending in `.` update the condition reg
+                outputs.extend(cr0_bits)
+
+                def eval_fn(s: NodeState, a: InstrArgs) -> None:
+                    target = a.reg_ref(0)
+                    val = cls.instrs_destination_first[mnemonic.rstrip(".")](a)
+                    target_val = s.set_reg(target, val)
+                    if target_val is None:
+                        target_val = val
+                    s.set_reg(
+                        Register("cr0_eq"),
+                        BinaryOp.icmp(
+                            target_val, "==", Literal(0, type=target_val.type)
+                        ),
+                    )
+                    # Use manual casts for cr0_gt/cr0_lt so that the type of target_val is not modified
+                    # until the resulting bit is .use()'d.
+                    target_s32 = Cast(
+                        target_val, reinterpret=True, silent=True, type=Type.s32()
+                    )
+                    s.set_reg(
+                        Register("cr0_gt"),
+                        BinaryOp(target_s32, ">", Literal(0), type=Type.s32()),
+                    )
+                    s.set_reg(
+                        Register("cr0_lt"),
+                        BinaryOp(target_s32, "<", Literal(0), type=Type.s32()),
+                    )
+                    s.set_reg(
+                        Register("cr0_so"),
+                        fn_op("MIPS2C_OVERFLOW", [target_val], type=Type.s32()),
+                    )
+
+            else:
+                eval_fn = lambda s, a: s.set_reg(
+                    a.reg_ref(0), cls.instrs_destination_first[mnemonic](a)
+                )
+        elif mnemonic in ("mtctr", "mtlr"):
             assert len(args) == 1 and isinstance(args[0], Register)
+            dest_reg = Register(mnemonic[2:])
             inputs = [args[0]]
-            outputs = [cls.instrs_implicit_destination[mnemonic][0]]
+            outputs = [dest_reg]
+            eval_fn = lambda s, a: s.set_reg(dest_reg, a.reg(0))
         elif mnemonic in cls.instrs_ppc_compare:
             assert len(args) == 3 and isinstance(args[1], Register)
             inputs = [r for r in args[1:] if isinstance(r, Register)]
             outputs = list(cr0_bits)
+
+            def eval_fn(s: NodeState, a: InstrArgs) -> None:
+                base_reg = a.reg_ref(0)
+                if base_reg != Register("cr0"):
+                    raise DecompFailure(
```

Comment:
pre-existing, but DecompFailure is a bit harsh, esp with the current error handling which a) is per block, and b) leaves stale registers, causing cascading failures

Reply:
Fixing this, unsurprisingly, changed the test output: functions that previously crashed with `DecompFailure` now partially decompile. (They're very messy because the non-CR0 bits are not supported by the branch instructions either)